HW4_Grigoriants/protein_tools.py

```diff
@@ -0,0 +1,308 @@
+    All nucleic acids (DNA and RNA) are showed in 5"-3" direction
+
+    Arguments:
+    - sequences (tuple[str] or list[str]): sequences to convert
+    - nucl_acids (str]: the nucleic acid that is prefered\n
+    Example: nucl_acids = "RNA" - convert to RNA\n
+                   nucl_acids = "DNA" - convert to DNA\n
+                   nucl_acids = "both" - convert to RNA and DNA
+    Return:
+    - dictionary: nucleic acids (str) as keys, collection of sequences (list) as values
+    """
+    rule_of_translation = sequences[0].maketrans(dictionaries.translation_rule)
+    rule_of_transcription = sequences[0].maketrans("AaUuCcGg", "TtAaGgCc")
+    nucl_acid_seqs = {"RNA": [], "DNA": []}
+    for sequence in sequences:
+        rna_seq = sequence.translate(rule_of_translation)
+        dna_seq = rna_seq.translate(rule_of_transcription)
+        if nucl_acids == "RNA":
+            nucl_acid_seqs["RNA"].append(rna_seq)
+            if sequence == sequences[-1]:
+                del nucl_acid_seqs["DNA"]
+        if nucl_acids == "DNA":
+            nucl_acid_seqs["DNA"].append(dna_seq)
+            if sequence == sequences[-1]:
+                del nucl_acid_seqs["RNA"]
+        if nucl_acids == "both":
+            nucl_acid_seqs["RNA"].append(rna_seq)
+            nucl_acid_seqs["DNA"].append(dna_seq)
+    return nucl_acid_seqs
+
+
+procedures_to_functions = {
+    "search_for_motifs": search_for_motifs,
+    "search_for_alt_frames": search_for_alt_frames,
+    "convert_to_nucl_acids": convert_to_nucl_acids,
+    "three_one_letter_code": three_one_letter_code,
+    "define_molecular_weight": define_molecular_weight,
+}
+
+
+def check_and_parse_user_input(
+    sequences: list[str] or tuple[str], **kwargs
+) -> dict and str:
+    """
+    Check if user input can be correctly processed\n
+    Parse sequences and arguments for desired procedure
+    
+    Arguments:
+    - sequences (list[str] or tuple[str]): sequences to process
+    - **kwargs - needed arguments for completion of desired procedure
+
+    Return:
+    - string: procedure name
+    - dictionary: a collection of procedure arguments and their values
+    """
+    if len(sequences) == 0:
+        raise ValueError("No sequences provided")
+    procedure = kwargs["procedure"]
+    if procedure not in procedures_to_functions.keys():
+        raise ValueError("Wrong procedure")
+    allowed_inputs = set(dictionaries.amino_acids.keys()).union(
+        set(dictionaries.amino_acids.values())
+    )
+    allowed_inputs.add("-")
+    if procedure != "three_one_letter_code":
+        allowed_inputs -= set(dictionaries.amino_acids.values())
+    for sequence in sequences:
+        allowed_inputs_seq = allowed_inputs.copy()
+        if procedure == "three_one_letter_code" and "-" in sequence:
+            allowed_inputs_seq -= set(dictionaries.amino_acids.keys())
+            if not all(
+                aminoacids in allowed_inputs_seq for aminoacids in sequence.split("-")
+            ):
```

Comment:
Т.к. allowed_inputs_seq является множеством, то не надо итерироваться по всей послед-ти и элементам множества, а достаточно использовать методы множеств:
```suggestion
            if not set(sequence.split("-")).issubset(allowed_inputs_seq):
```